aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import aws.sdk.kotlin.crt.Allocator
+import aws.sdk.kotlin.crt.CrtRuntimeException
+import aws.sdk.kotlin.crt.awsAssertOpSuccess
+import kotlinx.cinterop.*
+import libcrt.*
+
+/**
+ * A typealias to CRT's aws_<checksum_algorithm>_new functions (i.e aws_sha1_new, aws_sha256_new))
+ */
+internal typealias InitializeHashFn = (
+    allocator: CValuesRef<aws_allocator>?,
+) -> CPointer<aws_hash>?
+
+public open class ShaNative(
+    public val initializeFn: InitializeHashFn,
+) {
```

Comment:
**Question**: Why `public`?

Reply:
If it's not `public`, there are compiler errors because Sha1Native and Sha256Native are exposing the internal class

Reply:
Then we should find a way to hide those details no? It could be as simple as introducing an interface.

Reply:
Addressed by adding a `HashFunction` interface